enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,413 @@
+frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `ROUTER_SET_HTTP_HEADERS` are comma-delimited lists of base64 encoded values of
+the form `name:value` separated by comma.
+The template translates these values into an appropriate `http-response set-header` stanzas.
+For example, if `ROUTER_SET_HTTP_HEADERS` has encoded value `WC1GcmFtZS1PcHRpb25zOiBERU5Z,WC1YU1MtUHJvdGVjdGlvbjogMTttb2RlPWJsb2Nr` 
+of `X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block` then the template adds the following stanzas to
+the frontend fe_no_sni, frontend public, frontend fe_sni  section of the haproxy.config:
+```conf
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers via Route's Spec by Users/Devops/Developers.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `SetHTTPHeadersViaRoute`, which specify the headers to set in the
+backend section `Plain http backend or backend with TLS terminated at the edge or a secure backend with re-encryption.` of the
+haproxy.config at this [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609).
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `SetHTTPHeadersViaRoute` will be set in the same way as described in the section A) above for  `ROUTER_SET_HTTP_HEADERS`
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Do we need to set request headers ?
+- Do we need a delete option ?
+- Do we a need admission list option ?
+- Do we need to set load balance algo based on the header set ?
+- What shall be done if user sets a HSTS annotation in the route ? We can add a condition at [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)
+  to check if the annotation was set. If the annotation is set for hsts we will can use that value.
```

Comment:
In my opinion, we should prohibit specifying `strict-transport-security` using the new API, for the following reasons:

* HSTS is already [configurable through a route annotation](https://docs.openshift.com/container-platform/4.11/networking/routes/route-configuration.html#nw-enabling-hsts-per-route_route-configuration).
* We already have [an API for restricting HSTS in the cluster config](https://docs.openshift.com/container-platform/4.11/networking/routes/route-configuration.html#nw-enforcing-hsts-per-domain_route-configuration).
* The annotation value is already validated in the template (deads2k wrote an admission plugin, but we never merged it: https://github.com/openshift/kubernetes/pull/1216).
* Adding a second way to configure HSTS would significantly complicate this enhancement.

Reply:
Hi @Miciah !  we can restrict this header by putting a condition in the haproxy.template. WDYT ?
The other way will be adding a API field for the admission list of headers.
Yes apiserver code refers ingress config object and verifies the hts annotation set in the route.
>>The annotation value is already validated in the template (deads2k wrote an admission plugin, but we never >> merged it: https://github.com/openshift/kubernetes/pull/1216).
Which annotation ? I tried to search for hsts but did not find anything related to it in the commit.

Reply:
https://github.com/openshift/kubernetes/pull/1216  I think I can use it for verifying headers the ingress controller CR as per admission list I will set in the ingress config object.
For the verifying the headers mentioned in the route spec we can follow the similar approach we for hsts.

Reply:
> Hi @Miciah ! we can restrict this header by putting a condition in the haproxy.template. WDYT ? 

No, we can do better than that; we can restrict it in [route validation](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141), or write a new admission plugin alongside [the existing admission plugins](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route/apiserver/admission/).  

> The other way will be adding a API field for the admission list of headers. 

Why would we add an API field?  The list of prohibited headers is supposed to be an immutable list.  

> Yes apiserver code refers ingress config object and verifies the hts annotation set in the route.

Indeed, I found the correct link, and I have edited my earlier comment.  

> > > The annotation value is already validated in the template (deads2k wrote an admission plugin, but we
> > > never  merged it: [openshift/kubernetes#1216](https://github.com/openshift/kubernetes/pull/1216)).

Sorry, I mixed up PRs.  That admission plugin for the annotation was merged: https://github.com/openshift/openshift-apiserver/pull/224

> Which annotation ? I tried to search for hsts but did not find anything related to it in the commit.

`haproxy.router.openshift.io/hsts_header`.

Reply:
After discussing this with someone from the API team, I understand that adding validation for `spec.httpSetHeaders` to the existing route validation code in library-go is the way to go.  (The "admission list option" mentioned earlier in this EP might require an admission plugin though, depending on what you have in mind for that option.)

Reply:
@Miciah 
>> Why would we add an API field? The list of prohibited headers is supposed to be an immutable list.
I don't get it. We will need the input of headers to be denied either from the route Spec or route Annotation right ?

Reply:
i think I got it that we will be hard coding some headers in the library-go validation for routes.

Reply:
> I didn't get it regarding immutable lists. We will need the input of headers to be denied either from the route Spec or route Annotation right ?

No, "immutable" means that the user cannot change it.  

> i think I got it that we will be hard coding some headers in the library-go validation for routes.

Right, it can be hard-coded somewhere, and library-go is a reasonable place for it.

Reply:
done